File: robocode.api/src/main/java/net/sf/robocode/io/Logger.java
/**
 * Copyright (c) 2001-2019 Mathew A. Nelson and Robocode contributors
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * https://robocode.sourceforge.io/license/epl-v10.html
 */
package net.sf.robocode.io;


import net.sf.robocode.security.HiddenAccess;
import net.sf.robocode.util.StringUtil;
import robocode.control.events.BattleErrorEvent;
import robocode.control.events.BattleMessageEvent;
import robocode.control.events.IBattleListener;

import java.io.ByteArrayOutputStream;
import java.io.PrintStream;


/**
 * This is a class used for logging.
 *
 * @author Flemming N. Larsen (original)
 * @author Mathew A. Nelson (original)
 */
public class Logger {
	public static final PrintStream realOut = System.out;
	public static final PrintStream realErr = System.err;

	private static IBattleListener logListener;
	
	private final static StringBuffer logBuffer = new StringBuffer();

	public static void setLogListener(IBattleListener logListener) {
		Logger.logListener = logListener;
	}

	public static void logMessage(String s) {
		logMessage(s, true);
	}

	public static void logMessage(String s, boolean newline) {
		if (logListener == null) {
			if (System.getProperty("logMessages", "true").equalsIgnoreCase("true")) {
				s = StringUtil.toBasicLatin(s);
				if (newline) {
					realOut.println(s);
				} else {
					realOut.print(s);
					realOut.flush();
				}
			}
		} else {
			synchronized (logBuffer) {
				if (!HiddenAccess.isSafeThread()) {
					// we just queue it, to not let unsafe thread travel thru system
					logBuffer.append(s);
					logBuffer.append("\n");
				} else if (newline) {
					logListener.onBattleMessage(new BattleMessageEvent(logBuffer + s));
					logBuffer.setLength(0);
				} else {
					logBuffer.append(s);
				}
			}
		}
	}

	public static void logWarning(String s) {
		logMessage("Warning: " + s, true);
	}

	public static void logError(String message, Throwable t) {
		logError(message + ":\n" + toStackTraceString(t));
	}

	public static void logError(Throwable t) {
		logError(toStackTraceString(t));
	}

	public static void logError(String s) {
		if (logListener == null) {
			if (System.getProperty("logErrors", "true").equalsIgnoreCase("true")) {
				s = StringUtil.toBasicLatin(s);
				realErr.println(s);
			}
		} else {
			logListener.onBattleError(new BattleErrorEvent(s));
		}
	}

	private static String toStackTraceString(Throwable t) {
		if (t == null) {
			return "";
		}
		ByteArrayOutputStream baos = new ByteArrayOutputStream();
		PrintStream ps = new PrintStream(baos);

		t.printStackTrace(ps);
		ps.close();
		return baos.toString();
	}

	public static void printlnToRobotsConsole(String s) {
		// this will get redirected to robot's console
		System.out.println(s);
	}
}

File: plugins/dotnet/robocode.dotnet.api.net/src/net/sf/robocode/io/Logger.cs
using System;
using System.IO;
using System.Text;
using net.sf.robocode.security;
using robocode.net.sf.robocode.io;

namespace net.sf.robocode.io
{
    public class Logger
    {
        private static readonly StringBuilder logBuffer = new StringBuilder();
        public static readonly TextWriter realErr = Console.Error;
        public static readonly TextWriter realOut = Console.Out;

        private static ILogger logListener;

        public static void logError(Exception t)
        {
            logError(toStackTraceString(t));
        }

        public static void logError(string s)
        {
            if (logListener == null)
            {
                realErr.WriteLine(s);
            }
            else
            {
                logListener.logError(s);
            }
        }

        public static void logError(string message, Exception t)
        {
            var builder = new StringBuilder();
            builder.Append(message);
            builder.Append(":\n");
            builder.Append(toStackTraceString(t));
            logError(builder.ToString());
        }

        public static void logMessage(Exception e)
        {
            logMessage(toStackTraceString(e));
        }

        public static void logMessage(string s)
        {
            logMessage(s, true);
        }

        public static void logMessage(string s, bool newline)
        {
            if (logListener == null)
            {
                if (newline)
                {
                    realOut.WriteLine(s);
                }
                else
                {
                    realOut.Write(s);
                    realOut.Flush();
                }
                return;
            }
            lock (logBuffer)
            {
                if (!HiddenAccess.isSafeThread())
                {
                    logBuffer.Append(s);
                    logBuffer.Append("\n");
                }
                else if (newline)
                {
                    logBuffer.Append(s);
                    logListener.logMessage(logBuffer.ToString());
                    logBuffer.Length = 0;
                }
                else
                {
                    logBuffer.Append(s);
                }
            }
        }

        public static void logMessage(string message, Exception t)
        {
            var builder = new StringBuilder();
            builder.Append(message);
            builder.Append(":\n");
            builder.Append(toStackTraceString(t));
            logMessage(builder.ToString());
        }

        public static void printlnToRobotsConsole(string s)
        {
            //TODO System.@Out.println(s);
        }

        private static string toStackTraceString(Exception exception1)
        {
            return exception1.StackTrace;
        }
    }
}